### ISSUE TITLE
ENH: stats.logser: implement `cdf`/`sf`

### ISSUE THREAD

**Issue by ev-br** [scipy/scipy#3890]
Log-series distribution has a closed form expression for cdf in terms of an incomplete beta function, http://en.wikipedia.org/wiki/Logarithmic_distribution. 

Implementing could address  a cryptic comment,  

```
# FIXME: Fails _cdfvec
```

in https://github.com/scipy/scipy/blob/master/scipy/stats/_discrete_distns.py#L360.

A well hidden ticket, https://github.com/scipy/scipy/issues/1883 reporting a problem with `stats.logser.sf(np.inf, 0.5)`, can be related.

**Comment by chrisb83** [comment]
I just wanted to add the cdf which uses the incomplete beta function. however, it uses the "unregularized" beta function denoted B(x; a,b) on Wikipedia while `special.betainc` is the regularized beta function (denoted I_x(a, b) on Wikipedia which is a rescaled version. however, we need B(p; 0, k+1) to compute cdf(k, p) and the zero leads to NaN in betainc (we cannot rescale as`Gamma(0)` is not defined).

Any chance that the "unscaled" version can be added without too much trouble?

Derivation of the CDF: 
- `p_k = P(X = k) = c * p^k / k`
- `d/dp sum_j=1^k p**j / j = sum_j=1^k p**(j-1) = (1-p**k)/(1-p)`
- Integrating gives `sum_j=1^k p**j / j = log(1-p) - Integral(t**k (1-t)**(-1), t=0..p)`
and the last term can be expressed as the incomplete beta function B(p, 0, k+1). 

Note that integration with a = 0 is not a problem here as p < 1. But `betainc(p, 0, k)` is NaN.

**Comment by WarrenWeckesser** [comment]
We only need a special case of the incomplete beta function B(x; a, b), with a = k +1 and b = 0.  Plugging those values into the integral form of the beta function gives (using `Int` to denote the integal--I wish github had LaTeX markup!):

    B(p; k + 1, 0) = Int_0^p t^k / (1 - t) dt

According to Wolfram Alpha, that integral can be expressed in terms of the hypergeometric function ₂F₁:

    B(p; k + 1, 0) = Int_0^p t^k / (1 - t) dt = p^(k + 1) ₂F₁(1, k+1; k+2, p) / (k + 1)

We have ₂F₁ implemented as `scipy.special.hyp2f1`, so we can implement the CDF as

```
In [131]: def logser_cdf(k, p): 
     ...:     k = np.asarray(k) 
     ...:     p = np.asarray(p) 
     ...:     r = p**(k+1) * hyp2f1(1, k+1, k+2, p)/(k+1) 
     ...:     return 1 + r/np.log(1-p) 
     ...:                                                                                                                                                                                                                       
```
Compare the calculation to the existing implementation:
```
In [132]: logser_cdf([1, 2, 3, 4, 5], 0.93)                                                                                
Out[132]: array([0.34972135, 0.51234177, 0.61316644, 0.68349164, 0.73581359])

In [133]: from scipy.stats import logser                                                                                   

In [134]: logser.cdf([1, 2, 3, 4, 5], 0.93)                                                                                
Out[134]: array([0.34972135, 0.51234177, 0.61316644, 0.68349164, 0.73581359])
```
Looks good.

I'll create a PR with this implementation.

**Comment by WarrenWeckesser** [comment]
It turns out I won't create a PR.   `hyp2f1` has problems with large arguments.  So until we have a more robust implementation of  ₂F₁, we probably shouldn't use it for `logser._cdf`.

**Comment by mdhaber** [comment]
@steppi I imagined (then saw) that you might be interested in `hyp2f1`, so I thought I'd mention this issue. Do you think `hyp2f1` is in good enough shape now that it would make sense to implement `cdf`/`sf` methods based on it?

**Comment by steppi** [comment]
> @steppi I imagined (then saw) that you might be interested in `hyp2f1`, so I thought I'd mention this issue. Do you think `hyp2f1` is in good enough shape now that it would make sense to implement `cdf`/`sf` methods based on it?

Not yet. So far I've only touched the implementation for complex `z` and even there I haven't implemented the recurrences yet that let it work well with large arguments.

That being said, the special cases of  `hyp2f1` appearing in the incomplete beta function are better handled with a more specialized continued fraction expansion. I have an implementation in one of [my own projects](https://github.com/indralab/opaque/blob/584dcd35917ed26af2b09dc34969a4d5b2a94a97/opaque/stats/_stats.pyx#L44-L83) that could be ported over here with little effort. Although I've learned in just the past week from the [Boost documentation for Incomplete Beta functions](https://www.boost.org/doc/libs/1_38_0/libs/math/doc/sf_and_dist/html/math_toolkit/special/sf_beta/ibeta_function.html) that the continued fraction expansion I used isn't ideal.

**Comment by mdhaber** [comment]
I see. Speaking of Boost, is there anything there that could be used? It's relatively easy to add boost stuff to SciPy now.

**Comment by steppi** [comment]
> I see. Speaking of Boost, is there anything there that could be used? It's relatively easy to add boost stuff to SciPy now.

I just checked, and yes actually. Boost has an implementation of the unregularized incomplete beta function. See [here](https://www.boost.org/doc/libs/1_78_0/libs/math/doc/html/math_toolkit/sf_beta/ibeta_function.html).

**Comment by mdhaber** [comment]
@mckib2 Are these easy to include, too?

**Comment by mdhaber** [comment]
@steppi Is this still in the cards as part of gh-20223 or shall we just use `nsum` here? I confirmed that it's about as fast as what we have in the fastest cases, much faster in slower cases (slow convergence, vectorized calls), and we can get accurate `logcdf` and `logsf` from it.

**Comment by steppi** [comment]
> @steppi Is this still in the cards as part of [gh-20223](https://github.com/scipy/scipy/issues/20223) or shall we just use `nsum` here? I confirmed that it's about as fast as what we have in the fastest cases, much faster in slower cases (slow convergence, vectorized calls), and we can get accurate `logcdf` and `logsf` from it.

Yeah, it should be straightforward to just add the unregularized incomplete beta function from Boost.

**Comment by mdhaber** [comment]
@steppi Boost says the unregularized incomplete beta function requires $a > 0$ and $b > 0$, but we have $b = 0$.

**Comment by steppi** [comment]
> @steppi Boost says the unregularized incomplete beta function requires a > 0 and b > 0 , but we need b = 0 .

Ah Ok. I'd still like to try to tackle this, and will take a closer look in a week or two.

**Comment by mdhaber** [comment]
Happy ten year anniversary! I'm going to suggest the easiest thing: implement the unregularized beta by multiplying our regularized incomplete beta (`special.betainc`) by our complete beta (`special.beta`). The limitation that `b != 0` is not a practical problem because a tiny, nonzero `b` gives results that are about as accurate as we could hope for - about 13~14 digits over a broad range of parameters. 

<details>

```python3
import matplotlib.pyplot as plt
import numpy as np
from mpmath import mp
from scipy import special

mp.dps = 25

@np.vectorize
def sf_ref(k, p):
    return float(-mp.betainc(k, 0, 0, p) / mp.log(1 - p))

def sf(k, p):
    tiny = 1e-100
    return -special.betainc(k, tiny, p) * special.beta(k, tiny) / np.log1p(-p)

def relative_error(res, ref):
    with np.errstate(all='ignore'):
        return np.where(res != ref, abs(res - ref) / abs(ref), 0)

p = np.linspace(0.01, 0.99, 25)
k = np.linspace(1, 5001, 25)[:, np.newaxis]
res = sf(k, p)
ref = sf_ref(k, p)
err = relative_error(res, ref)
logerr = np.log10(err)
logerr[logerr < -16] = -16

plt.contourf(p, k.ravel(), logerr, cmap='viridis')
plt.xlabel("p")
plt.ylabel("k")
plt.title("log10 error")
plt.colorbar()
plt.show()
```

</details>

![image](https://github.com/user-attachments/assets/a36925a2-e1f3-4d69-930d-8cba317ad20e)

That yellow strip is the survival function jumping from some extremely tiny number (e.g. `1e-200`) to zero a little before the reference implementation (using mpmath). To the upper left of the yellow, the error is exactly zero because the reference value is too small to represent as a nonzero number in `float64`. To the lower right of the yellow, the relative error is typically ~`1e-13`.

For $p = 0.001$:

<details>

```python3
default_colors = plt.rcParams['axes.prop_cycle'].by_key()['color']
c1 = default_colors[0]
c2 = default_colors[2]

p = 0.001
k = np.arange(1, 110)
res = sf(k, p)
ref = sf_ref(k, p)
err = relative_error(res, ref)
plt.semilogy(k, abs(res - ref), color=c1)
plt.ylabel('absolute error', color=c1)

plt.twinx()

plt.semilogy(k, err, color=c2)
plt.xlabel('k')
plt.ylabel('relative error', color=c2)
plt.title('Error of proposed SF implementation')

```

</details>

![image](https://github.com/user-attachments/assets/3e3d1ced-9936-4f1d-b955-84aeb350f8d7)

And on the other side, for $p = 0.999$:
![image](https://github.com/user-attachments/assets/b4ef7cec-0f83-4130-8ea8-dc6a3b8a6d6b)

About the same story is true for for $10^{-15} < p < 1 - 10^{-15}$, AFAICT.

And we might be able to do just a little better now that it's so easy to work with Boost: use its regularized beta incomplete directly - with small `b`, or try removing the limitation that `b != 0`. My guess is that whatever technique it's using to implement the function converges for `b = 0` just fine. The typical continued fraction ([DLMF 8.17v](https://dlmf.nist.gov/8.17)) does, even though it doesn't advertise it.

It's not perfect, but it's better than many of our distributions, certainly better than what we have now, and probably not worth waiting another ten years for something better : )
